### DISCUSSION TITLE
Dropping root `package.json` version

### DISCUSSION THREAD

**Discussion by overbit** [amplication/amplication#6789] [Ideas] [answered]
## TL;DR;

The proposal is to remove the version from the root `package.json` and the process in place to bump it on release, and update the `amplication-client` to **rely at build-time  or run-time exclusively on the github release version** as per amplication website.

## Details

Currently we rely on the `package.json` version to understand what version of amplication we are running. 
This value, is used during the client build process to save the version that will be shown in the client interface. 

As amplication nature is not monolithic but services based, this unique version is not ideal.
Moreover, the process of updating `package.json` version, makes the version of Nx, currently used in amplication, flagging all the projects in the monorepo as affected by a change, and as a consequence,  our deployment strategy will process all projects as changed even if we only touched, i.e. `amplication-server`.

**Comment by yuval-hazaz** [comment]
@overbit I think we should split the discussion to two parts:
1. How we want to manage the versions on Amplication with nx, versioning strategy, reason to change the current state, etc
2. Once we agree on that, we can look into how to achieve that, what to do with the client, and what to do with the package.json

**Reply by overbit** [reply]
To articulate a bit more, at the moment, the currently used version of nx consider changes like the one in devDeps, deps and version in package.json  to affect all the projects in the monorepo making impossible to use nx affected commands effectively.

In order to release DSG versions only when it's affected, without a big workaround that will involve checking old tagged dsg and only create new versions without retagging, we will need to avoid updating the version unless we need to affects ALL projects

**Reply by overbit** [reply]
To support DSG independent versioning we have then the following options:

Option 1:
keep root package.json version
1. move `data-service-generator` out of nx workspace
2. update all `data-service-generator` dependencies 
3. maintain `data-service-generator` dependencies manually
4. create GitHub action specific to `data-service-generator` (as it was in a different repository)
5. create e2e GitHub action specific to `data-service-generator` (as it was in a different repository)
6. update current e2e GitHub actions to filter out `data-service-generator` commands
7. migrate all `data-service-generator` nx scripts to either native or custom script (where native are not available)
8. publish and maintain published version of common library used by `data-service-generator` as `@amplication/util/logging`
9. create scripts to understand when `data-service-generator` is affected by a change

Option 2:
keep root package.json version
1. keep `data-service-generator` in nx workspace
2. create GitHub action specific to `data-service-generator` (as it was in a different repository)
3. update current GitHub actions to filter out `data-service-generator` commands
4. create e2e GitHub action specific to `data-service-generator` (as it was in a different repository)
5. update current e2e GitHub actions to filter out `data-service-generator` commands

Option 3:
removing version from package.json
1. update `amplication-client` script to read from GitHub latest release tag

**Reply by yuval-hazaz** [reply]
in any of the above options, you didn't really explain how to maintain the version numbers. is that going to be a pckage.json per each project? how will you bunp it? what will be the impact?

**Reply by overbit** [reply]
@yuval-hazaz good point. 
There will not be any version defined in the source code in per se, but we will continue to version services relying on the GitHub Releases, tagging images using the Release tag.   

This will change the way we release as we won't need a manual version bump PR before release

**Reply by abrl91** [reply]
I vote for option 3

**Reply by yuval-hazaz** [reply]
so, in other words, option 3 says - Keep using nx as usual for everything, but stop using it for version numbering. 
Instead of counting on the version number from package.json we will use releases and tags on GitHub. 

Because we do that, we will be missing the ability to show the version number in the UI - and to workaround it, we will be fetching the version from GitHub directly, assuming that the latest version on GitHub is always live on prod. if that is what you mean, I don't think it is a good solution for various reasons.  Or do you suggest something else?

**Reply by yuval-hazaz** [reply]
to be precise - I think the solution itself is good. I don't think the solution for the client issue (the way I understand it) is good

**Reply by overbit** [reply]
> so, in other words, option 3 says - Keep using nx as usual for everything, but stop using it for version numbering.
Instead of counting on the version number from package.json we will use releases and tags on GitHub.
Because we do that, we will be missing the ability to show the version number in the UI - and to workaround it, we will be fetching the version from GitHub directly, assuming that the latest version on GitHub is always live on prod. if that is what you mean, I don't think it is a good solution for various reasons. Or do you suggest something else?

Nx was never the owner of versioning since it **doesn't** need or care about that package.json version to calculate the affected applications.

For the client we will collect it at build time so there is not runtime dependency on GitHub from this point of view.
Here is what I mean in code:
https://github.com/amplication/amplication/pull/6787/files#diff-9634303137442d185bc523324a75aef222e748c772138aacc82e00e1f8d3ba10R17

**Reply by mulygottlieb** [reply]
@yuval-hazaz what is bothering you regarding the last part about how to show the version number in the client? Would you like to suggest a different approach?
Remember that we do want (IMO) to show one unified version to the user in the UI, that corresponds with a GH release (and release notes), and GH published release seems right. 
@overbit I don't understand how this can be not a runtime dependency - if we release a new version with no client changes, the client doesn't get rebuilt, so how will it get the new version number?

**Reply by overbit** [reply]
We will also be able to make `version.ts` part of the repository and use a manual workflow to update them and create a tag that will be used in production releases. I'll prepare an example with a diagram to showcase the idea.

**Reply by overbit** [reply, marked as answer]
Here the solution with a video showing the new flow:
https://github.com/amplication/amplication/pull/6787
